src/cargo/util/auth/mod.rs

```diff
@@ -88,17 +88,29 @@ fn credential_provider(config: &Config, sid: &SourceId) -> CargoResult<Vec<Vec<S
             vec![vec!["cargo:token".to_string()]]
         }
     };
+    let key = "registry.global-credential-providers";
+    let env_key = ConfigKey::from_str(key).as_env_key().to_string();
     let global_providers = config
-        .get::<Option<Vec<Value<String>>>>("registry.global-credential-providers")?
+        .get::<Option<Vec<Value<String>>>>(key)?
         .filter(|p| !p.is_empty() && allow_cred_proc)
         .map(|p| {
-            p.iter()
-                .rev()
+            // Cargo normally merges configuration from environment variables at the end of a list,
+            // but here we want to the global credential provider attempted first.
+            let env_iter = config.get_env(&env_key).ok().map(|val| Value {
+                val,
+                definition: Definition::Environment(env_key),
+            });
+            let cfg_iter = p
+                .iter()
+                .filter(|p| !matches!(p.definition, Definition::Environment(_)));
```

Comment:
This doesn't look correct, no?  `cargo --config` takes precedence over other stuff.

Reply:
I'd forgotten about `--config`. After adding tests, I found the current behavior to be even more unexpected, so I filed #12506

This PR will probably need to wait until we resolve that.

Reply:
Oh no. #12506 doesn't looks too nice…